src/debtManagers/HookAMM.sol

```diff
@@ -0,0 +1,238 @@
+        maxDolaPrice = _maxDolaPrice;
+    }
+
+    function setMinDolaPrice(uint _minDolaPrice) external onlyGov {
+        require(maxDolaPrice >= _minDolaPrice, "Max DOLA price must be higher or equal to min DOLA price");
+        minDolaPrice = _minDolaPrice;
+    }
+
+    function getVirtualDolaReserves() internal view returns(uint) {
+        return sqrt(1e18 * getK() / minDolaPrice);
+    }
+
+    /**
+     * @dev Allows users to buy DBR with DOLA.
+     * WARNING: Never expose this directly to a UI as it's likely to cause a loss unless a transaction is executed immediately.
+     * Instead use the sDolaHelper function or custom smart contract code.
+     * @param exactDolaIn The exact amount of DOLA to spend.
+     * @param exactDbrOut The exact amount of DBR to receive.
+     * @param to The address that will receive the DBR.
+     */
+    function buyDBR(uint exactDolaIn, uint exactDbrOut, address to) external buyHook {
+        require(to != address(0), "Zero address");
+        _invariantCheck(exactDolaIn, exactDbrOut, getDbrReserve());
+        dola.transferFrom(msg.sender, address(this), exactDolaIn);
+        uint dbrBal = dbr.balanceOf(address(this));
+        if(exactDbrOut > dbrBal)
+            dbr.mint(address(this), exactDbrOut - dbrBal);
+        uint exactDbrOutAfterFee = exactDbrOut * (1e18 - dbrBuyFee) / 1e18;
+        feesAccrued += exactDbrOut - exactDbrOutAfterFee;
+        dbr.transfer(to, exactDbrOutAfterFee);
+        emit BuyDBR(msg.sender, to, exactDolaIn, exactDbrOut);
+    }
+
+    function burnDBR(uint exactDolaIn, uint exactDbrBurn) external buyHook {
+        _invariantCheck(exactDolaIn, exactDbrBurn, getDbrReserve());
+        dola.transferFrom(msg.sender, address(this), exactDolaIn);
+        dbr.burn(exactDbrBurn);
+        emit Burn(msg.sender, exactDolaIn, exactDbrBurn);
+    }
+
+    function buyDola(uint exactDbrIn, uint exactDolaOut, address to) external buyHook {
+        require(to != address(0), "Zero address");
+        uint k = getK();
+        uint dolaReserve = getDolaReserve();
+        uint dbrReserve = k / dolaReserve;
+        _invariantCheck(dbrReserve - exactDbrIn, dolaReserve - exactDolaOut, getDolaReserve());
+        dbr.transferFrom(msg.sender, address(this), exactDolaOut);
+        dola.transfer(to, exactDbrIn);
+        emit BuyDOLA(msg.sender, to, exactDbrIn, exactDolaOut);
+    }
+
+    function _invariantCheck(uint newDolaReserve, uint newDbrReserve, uint k) internal view {
+        if(newDolaReserve * newDbrReserve < k){
+            if(1e18 * newDbrReserve / newDolaReserve > maxDolaPrice){
+                uint maxDolaReserve = sqrt(1e18 * k / maxDolaPrice);
+                uint maxDbr = k / maxDolaReserve;
+                uint excessDola = newDolaReserve - maxDolaReserve;
+                uint excessDbr = newDbrReserve - maxDbr;
+                if(1e18 * excessDbr / excessDola >= maxDolaPrice) revert Invariant();
```

Comment:
Why are we enforcing price invariants when the price is determined by our own AMM? This will essentially halt trading beyond price constraints while the intended behavior would be to continue trading at the price constraints. I think maybe the invariant design might be inappropriate here and instead we could opt for a simple price function that derives price from xy=k up until the constraints then returns fixed prices beyond them.